x/incentive/keeper/rewards_earn_accum_test.go

```diff
@@ -86,11 +90,93 @@ func (suite *AccumulateEarnRewardsTests) TestStateUpdatedWhenBlockTimeHasIncreas
 	})
 }
 
+func (suite *AccumulateEarnRewardsTests) TestStateUpdatedWhenBlockTimeHasIncreased_bkava() {
+	vaultDenom1 := "bkava-meow"
+	vaultDenom2 := "bkava-woof"
+
+	earnKeeper := newFakeEarnKeeper().
+		addVault(vaultDenom1, earntypes.NewVaultShare(vaultDenom1, d("800000"))).
+		addVault(vaultDenom2, earntypes.NewVaultShare(vaultDenom2, d("200000")))
+
+	liquidKeeper := newFakeLiquidKeeper().
+		addDerivative(vaultDenom1, i(800000)).
+		addDerivative(vaultDenom2, i(200000))
+
+	suite.keeper = suite.NewKeeper(&fakeParamSubspace{}, nil, nil, nil, nil, nil, nil, nil, liquidKeeper, earnKeeper)
+
+	globalIndexes := types.MultiRewardIndexes{
+		{
+			CollateralType: vaultDenom1,
+			RewardIndexes: types.RewardIndexes{
+				{
+					CollateralType: "earn",
+					RewardFactor:   d("0.02"),
+				},
+				{
+					CollateralType: "ukava",
+					RewardFactor:   d("0.04"),
+				},
+			},
+		},
+		{
+			CollateralType: vaultDenom2,
+			RewardIndexes: types.RewardIndexes{
+				{
+					CollateralType: "earn",
+					RewardFactor:   d("0.02"),
+				},
+				{
+					CollateralType: "ukava",
+					RewardFactor:   d("0.04"),
+				},
+			},
+		},
+	}
+
+	suite.storeGlobalEarnIndexes(globalIndexes)
+	previousAccrualTime := time.Date(1998, 1, 1, 0, 0, 0, 0, time.UTC)
+	suite.keeper.SetEarnRewardAccrualTime(suite.ctx, vaultDenom1, previousAccrualTime)
+	suite.keeper.SetEarnRewardAccrualTime(suite.ctx, vaultDenom2, previousAccrualTime)
+
+	newAccrualTime := previousAccrualTime.Add(1 * time.Hour)
+	suite.ctx = suite.ctx.WithBlockTime(newAccrualTime)
+
+	rewardPeriod := types.NewMultiRewardPeriod(
+		true,
+		"bkava",         // reward period is set for "bkava" to apply to all vaults
+		time.Unix(0, 0), // ensure the test is within start and end times
+		distantFuture,
+		cs(c("earn", 2000), c("ukava", 1000)), // same denoms as in global indexes
+	)
+	suite.keeper.AccumulateEarnRewards(suite.ctx, rewardPeriod)
+
+	// check time and factors
+
+	suite.storedTimeEquals(vaultDenom1, newAccrualTime)
+	suite.storedTimeEquals(vaultDenom2, newAccrualTime)
+
+	// Each vault gets the same ukava per second, assuming shares prices are the same.
+	// The share amount determines how much is actually distributed to the vault.
+	expectedIndexes := types.RewardIndexes{
+		{
+			CollateralType: "earn",
+			RewardFactor:   d("7.22"),
+		},
+		{
+			CollateralType: "ukava",
+			RewardFactor:   d("3.64"),
+		},
+	}
```

Comment:
This part is a bit confusing with multiple bkava vaults, different bkava vaults have the same per second per share saved in indexes assuming vault shares are 1:1 to bkava-xx supply.

For example with the following vaults with 100ukava per second set for "bkava"
* 95,000bkava-1
  * 95% of 100ukava per second == 95ukava per second
  * 95 / 95_000 == 0.001 per second per share
* 5,000bkava-2
  * 5% of 100ukava per second == 5ukava per second
  * 5 / 5_000 == 0.001 per second per share

After 1 hour:

  * rewards per share == 3600 seconds * 0.001ukava per second == 3.6ukava
  * bkava-1 vault
    * total rewards = 3.6ukava * 95,000 == 342000ukava
  * bkava-2 vault
    * total rewards = 3.6ukava * 5,000 == 18000ukava

  * total distributed to all bkava vaults == 360000ukava
  * 360000ukava / 3600seconds == 100ukava per second

Reply:
interesting